Fix #1, Update to dev guide, add table, add CRC output

avan989 opened the pull request:
**Describe the contribution**
Updated to dev. guide, added sample table, added CRC output

**Testing performed**
Steps taken to test the contribution:
1. Build 
2. Send Ground Command
3. Verify Results

```
EVS Port1 42/1/SAMPLE_APP 3: SAMPLE: NOOP command
EVS Port1 42/1/SAMPLE_APP 4: SAMPLE: RESET command
1980-012-14:03:25.00026 Sample App: Table Value 1: 1  Value 2: 2
1980-012-14:03:25.00161 Sample App: CRC = 0xFFFF9C00
```

output from tblCRCTool:

```
Table File Name:            ../../exe/cpu1/cf/sample_table.tbl
Table Size:                 4 Bytes
Expected TS Validation CRC: 0xFFFF9C00
```

**Expected behavior changes**
no impact to behavior

**System(s) tested on:**
 - Hardware
 - OS: Ubuntu 18.04.03
 - Versions CFS 6.7

**Contributor Info**
Anh Van, NASA Goddard

**Community contributors**
You must attach a signed CLA (required for acceptance) or reference one already submitted

skliper commented on CMakeLists.txt:
```diff
@@ -1,10 +1,12 @@
 cmake_minimum_required(VERSION 2.6.4)
 project(CFE_SAMPLE_APP C)
 
+include_directories(../../cfe/fsw/cfe-core/src/tbl)
```
This should probably be:
    include_directories(${cfe-core_MISSION_DIR}/src/tbl)

skliper commented on fsw/src/sample_app.c:
```diff
@@ -26,65 +26,86 @@
-#include "sample_app_msgids.h"
-#include "sample_app_msg.h"
-#include "sample_app_events.h"
-#include "sample_app_version.h"
```
I'd expect these to be here instead of sample_app.h.  Is this what the app developers guide recommends?

skliper commented on fsw/src/sample_app.c:
```diff
@@ -170,52 +285,112 @@ void SAMPLE_ProcessCommandPacket(void)
+/* NoopCmd -- SAMPLE NOOP commands                                            */
+/*                                                                            */
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * **/
+void NoopCmd( CFE_SB_MsgPtr_t msg )
```
SAMPLE_NoopCmd recommended

skliper commented on fsw/src/sample_app.c:
```diff
@@ -224,26 +399,70 @@ void SAMPLE_ReportHousekeeping(void)
-    bool result = true;
+boolean SAMPLE_VerifyCmdLength( CFE_SB_MsgPtr_t msg, uint16 ExpectedLength )
+{
+    boolean result = TRUE;
```
If you compile with deprecated elements removed this will fail.  Utilize bool/true/false in code.

skliper commented on fsw/src/sample_app.c:
```diff
@@ -255,14 +474,137 @@ bool SAMPLE_VerifyCmdLength(CFE_SB_MsgPtr_t msg, uint16 ExpectedLength)
+                          ActualLength,
+                          ExpectedLength);
+
+        result = FALSE;
```
false

avan989 replied:
see new commit for update

skliper commented on fsw/src/sample_app.c:
```diff
@@ -255,14 +474,137 @@ bool SAMPLE_VerifyCmdLength(CFE_SB_MsgPtr_t msg, uint16 ExpectedLength)
+/*                                                                 */
+/*                                                                 */
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+void Get_Crc(const char *fileName)
```
SAMPLE_GetCRC

avan989 replied:
see new commit for update

skliper commented on fsw/src/sample_app.c:
```diff
@@ -224,26 +399,70 @@ void SAMPLE_ReportHousekeeping(void)
+/*         This function Process Ground Station Command                       */
+/*                                                                            */
+/* * * * * * * * * * * * * * * * * * * * * * * *  * * * * * * *  * *  * * * * */
+void  SAMPLE_Process_CC(CFE_SB_MsgPtr_t msg)
```
SAMPLE_ProcessCC

avan989 replied:
see new commit for update

skliper commented on fsw/src/sample_app.h:
```diff
@@ -37,25 +37,42 @@
 #include "sample_app_msgids.h"
 #include "sample_app_msg.h"
-#include "sample_app_events.h"
-#include "sample_app_version.h"
```
*_events.h and *_version.h follow the app development model, I recommend against removing them.  Typically define these values in individual header files such that changes to logic are more obvious (vs changes to id's or version numbers).

skliper replied:
Resolved, follows app dev guide model

skliper commented on fsw/src/sample_app.c:
```diff
@@ -368,10 +368,10 @@ void SAMPLE_ReportHousekeeping( CFE_SB_MsgPtr_t msg )
+void SAMPLE_NoopCmd( CFE_SB_MsgPtr_t msg )
 {
     uint16 ExpectedLength = sizeof(SAMPLE_NoArgsCmd_t);
 
```
Report version number in noop message.

avan989 commented on fsw/src/sample_app.h:
```diff
@@ -37,18 +37,70 @@
+#include "sample_app_msgids.h"
+#include "sample_app_msg.h"
+#include "sample_app_events.h"
+#include "sample_app_version.h"
```
removed events, version and placed in sample_app's header. sample_app_msgids.h and perfids.h are part of the developer guide. I recommend keeping sample_app_msg.h (the command code and structure) separate.  There is an inconsistency with how the developer guide implement the code and what is needed for the groundsystem tool to auto generate the ground command. It'll make the process a little messy. See new commit for changes

skliper replied:
Sorry, I should have explained better... I just meant they typically aren't included from the sample_app.h file, I'd expect them to be included from the c file.

skliper replied:
Resolved, follows app dev guide model

avan989 commented on CMakeLists.txt:
```diff
@@ -1,10 +1,12 @@
 cmake_minimum_required(VERSION 2.6.4)
 project(CFE_SAMPLE_APP C)
 
+include_directories(../../cfe/fsw/cfe-core/src/tbl)
```
Needed the src/tbl for a manual CRC calculation that was done. See new commit for changes.

jphickey replied:
The code in "src" is supposed to be private to that module.  Applications should only use the public CFE API exposed in `cfe/fsw/cfe-core/src/inc`.

skliper replied:
Resolved, no longer includes internal headers

avan989 commented on fsw/src/sample_app.h:
```diff
@@ -37,18 +37,70 @@
 #include "cfe_es.h"
+#include "cfe_tbl_filedef.h"
+#include "cfe_tbl_task.h"
+#include "cfe_tbl_internal.h"
```
Needed CFE_TBL_RegistryRec_t and CFE_TBL_LoadBuff_t from cfe_tbl_task.h 
Needed the variable CFE_TBL_NOT_FOUND in cfe_tbl_internal.

I don't think they are critical unless you want to manually do an CRC calculation instead of relying on CFS. See new commit for changes.

skliper replied:
cfe_tbl_internal.h is intented to be internal to tbl use only, what depends on this definition?

jphickey replied:
I concur that the `cfe_tbl_task.h` and `cfe_tbl_internal.h` files should _not_ be required here.  These are internal to the TBL subsystem implementation and not part of its public API.  As far as I know, no other table-using CFS apps need to access these internal structures.  With the exception of elf2cfetbl, which documented as needing cleanup to remove this.

avan989 commented:
new updates, fixed dependency on ../src/tbl. Clean up code and tested. 
Results are consistence with CRC tools:
```
EVS Port1 42/1/SAMPLE_APP 3: SAMPLE: NOOP command
EVS Port1 42/1/SAMPLE_APP 4: SAMPLE: RESET command
1980-012-14:04:49.00100 Sample App: Table Value 1: 1  Value 2: 2
1980-012-14:04:49.00102 Sample App: CRC: 0xFFFF9C00
```

avan989 commented:
Combine commit for review

avan989 commented:
Test result:
```
EVS Port1 42/1/SAMPLE_APP 3: SAMPLE: NOOP command  Version 1.1.0.0
EVS Port1 42/1/SAMPLE_APP 4: SAMPLE: RESET command
1980-012-14:03:28.50036 Sample App: Table Value 1: 1  Value 2: 2
1980-012-14:03:28.50037 Sample App: CRC: 0xFFFF9C00
```

avan989 commented:
1) updated to match cfe for command and housekeeping. 
2) That was an leftover from an copy and paste. Fixed. 
3) Can you clarify about the commit log. The only commit in my log is this current one.
